packages/ra-core/src/controller/input/ReferenceInputController.tsx

```diff
@@ -147,183 +128,63 @@ interface State {
-        const id = input.value;
-        if (id) {
-            crudGetManyAccumulate(reference, [id]);
-        }
-    };
-
-    fetchOptions = (props = this.props) => {
-        const {
-            crudGetMatchingAccumulate,
-            filter: filterFromProps,
-            reference,
-            referenceSource,
-            resource,
-            source,
-        } = props;
-        const { pagination, sort, filter } = this.state;
-
-        crudGetMatchingAccumulate(
-            reference,
-            referenceSource(resource, source),
-            pagination,
-            sort,
-            {
-                ...filterFromProps,
-                ...filter,
-            }
-        );
-    };
-
-    fetchReferenceAndOptions(props) {
-        this.fetchReference(props);
-        this.fetchOptions(props);
-    }
-
-    render() {
-        const {
-            input,
-            referenceRecord,
-            matchingReferences,
-            onChange,
-            children,
-            translate,
-        } = this.props;
-        const { pagination, sort, filter } = this.state;
-
-        const dataStatus = getDataStatus({
-            input,
-            matchingReferences,
-            referenceRecord,
-            translate,
-        });
-
-        return children({
-            choices: dataStatus.choices,
-            error: dataStatus.error,
-            isLoading: dataStatus.waiting,
-            onChange,
-            filter,
-            setFilter: this.debouncedSetFilter,
-            pagination,
-            setPagination: this.setPagination,
-            sort,
-            setSort: this.setSort,
-            warning: dataStatus.warning,
-        });
-    }
-}
-
-const makeMapStateToProps = () =>
-    createSelector(
-        [
-            getReferenceResource,
-            getPossibleReferenceValues,
-            (_, props) => props.input.value,
-        ],
-        (referenceState, possibleValues, inputId) => ({
-            matchingReferences: getPossibleReferences(
-                referenceState,
-                possibleValues,
-                [inputId]
-            ),
-            referenceRecord: referenceState && referenceState.data[inputId],
-        })
-    );
-
-const ReferenceInputController = compose(
-    withTranslate,
-    connect(
-        makeMapStateToProps(),
-        {
-            crudGetManyAccumulate: crudGetManyAccumulateAction,
-            crudGetMatchingAccumulate: crudGetMatchingAccumulateAction,
-        }
-    )
-)(UnconnectedReferenceInputController);
-
-ReferenceInputController.defaultProps = {
-    referenceSource: defaultReferenceSource, // used in makeMapStateToProps
+export const ReferenceInputController: FunctionComponent<Props> = ({
+    input,
+    onChange,
+    children,
+    perPage = 25,
+    filter: permanentFilter = {},
+    reference,
+    filterToQuery,
+    referenceSource = defaultReferenceSource,
+    resource,
+    source,
+}) => {
+    const translate = useTranslate();
```

Comment:
why don't you extract the body of that component to a `useReferenceInput` hook? You've already done 90% of the job.